src/Experimental/RAGTools/types.jl

```diff
@@ -134,6 +135,61 @@ chunkdata(index::ChunkEmbeddingsIndex) = embeddings(index)
 
 # For backward compatibility
 const ChunkIndex = ChunkEmbeddingsIndex
+
+# TODO: where to put these?
+indexid(index::AbstractManagedIndex) = index.id
+chunks(index::AbstractManagedIndex) = index.chunks
+sources(index::AbstractManagedIndex) = index.sources
+
+# TODO: what about this?
+using Pinecone: Pinecone, PineconeContextv3, PineconeIndexv3
+
+"""
+    PineconeIndex
+
+Main struct for storing document chunks and their embeddings along with the necessary Pinecone context for connecting to Pinecone.
+
+# Fields
+- `id::Symbol`: unique identifier of each index (a symbol of the Pinecone index namespace)
+- `pinecone_context::Pinecone.PineconeContextv3`: Pinecone API key
+- `pinecone_index::Pinecone.PineconeIndexv3`: Pinecone index
+- `pinecone_namespace::String`: name of the namespace inside the Pinecone index
+- `chunks::Vector{<:AbstractString}`: underlying document chunks / snippets
+- `embeddings::Union{Nothing, Matrix{<:Real}}`: for semantic search
+- `tags::Union{Nothing, AbstractMatrix{<:Bool}}`: for exact search, filtering, etc. This is often a sparse matrix indicating which chunks have the given `tag` (see `tag_vocab` for the position lookup)
+- `tags_vocab::Union{Nothing, Vector{<:AbstractString}}`: vocabulary for the `tags` matrix (each column in `tags` is one item in `tags_vocab` and rows are the chunks)
+- `sources::Vector{<:AbstractString}`: sources of the chunks
+- `metadata::Vector{Dict{String, Any}}`: metadata for each chunk/embedding stored in Pinecone
+"""
+@kwdef struct PineconeIndex{
+    T1 <: Union{Nothing, AbstractString},
+    T2 <: Union{Nothing, Matrix{<:Real}},
+    T3 <: Union{Nothing, AbstractMatrix{<:Bool}}
+} <: AbstractManagedIndex
+    # TODO: id should be a combination of index + namespace?
+    id::Symbol  # namespace
+    # TODO: these should not be v3, maybe?
+    pinecone_context::Pinecone.PineconeContextv3
+    pinecone_index::Pinecone.PineconeIndexv3
+    pinecone_namespace::String
+    # underlying document chunks / snippets
+    chunks::Vector{T1} = nothing
+    # for semantic search
+    embeddings::T2 = nothing
+    # for exact search, filtering, etc.
+    # expected to be some sparse structure, eg, sparse matrix or nothing
+    # column oriented, ie, each column is one item in `tags_vocab` and rows are the chunks
+    tags::T3 = nothing
+    tags_vocab::Union{Nothing, Vector{<:AbstractString}} = nothing
+    sources::Union{Nothing, Vector{<:AbstractString}} = nothing
+    # metadata for each chunk
+    # TODO: should be changed to `extras`? but different type -- this needs to be vector of dicts
+    metadata::Vector{Dict{String, Any}} = Vector{Dict{String, Any}}()
+end
+HasKeywords(::PineconeIndex) = false
+HasEmbeddings(::PineconeIndex) = true
+embeddings(index::PineconeIndex) = index.embeddings
```

Comment:
should we have these accessors and fields if we know that the data isn’t there? we hold the data in the individual candidatechunk objects

Reply:
If these are removed, then getting the embeddings in `retrieve` should be changed.